File: Exploratory/AtomicityDec3/src/a_package/controllerImpl.c
// Variables and constants
int a_controllerImpl = 0;
int used_a_controllerImpl = 0;

// Subroutines
void controllerImpl_controller_fmiDoStep() {
// [Internal] This will block until the guard has been met (no nice way of doing this in OpenMP 2.5)
	BOOL completed = FALSE; // [Internal] Private to the thread by default
// [Internal] Try and get lock
#pragma omp critical (CONTROLLERIMPL)
	{
// Translated code
		used_a_controllerImpl = ((a_controllerImpl) / 2);
// [Internal] Set completed flag
		completed = TRUE;
	}
}

void controllerImpl_fmiSetInteger(int p) {
// [Internal] This will block until the guard has been met (no nice way of doing this in OpenMP 2.5)
	BOOL completed = FALSE; // [Internal] Private to the thread by default
// [Internal] Try and get lock
#pragma omp critical (CONTROLLERIMPL)
	{
// Translated code
		a_controllerImpl = p;
// [Internal] Set completed flag
		completed = TRUE;
	}
}

File: Exploratory/AtomicityDec3/src/atomicityDec3_java/controllerImpl.java
package atomicityDec3_java;
// Protected: controllerImpl

public class controllerImpl
{
	
	// Instance variables and constants
	protected static int a = 0;
	protected static int used_a = 0;
	private static Object lock = new Object();
	
	public controllerImpl()
	{
		
	}
	
	// Subroutines
	public void controller_fmiDoStep()
	{
		// [Internal] This will block until the guard has been met
		boolean completed = false; 	 // private to the thread by default
		// [Internal] Try and get lock
		synchronized(lock)
		{
			// Translated code
			used_a = ((a) / 2);
			// [Internal] Set completed flag
			completed = true;
		}
	}
	
	public void fmiSetInteger(int p)
	{
		// [Internal] This will block until the guard has been met
		boolean completed = false; 	 // private to the thread by default
		// [Internal] Try and get lock
		synchronized(lock)
		{
			// Translated code
			a = p;
			// [Internal] Set completed flag
			completed = true;
		}
	}
	
	public void fmiDoStep()
	{
		// [Internal] This will block until the guard has been met
		boolean completed = false; 	 // private to the thread by default
		// [Internal] Try and get lock
		synchronized(lock)
		{
			// Translated code
			used_a = ((a) / 2);
			System.out.println("used_a: " + used_a);
			// [Internal] Set completed flag
			completed = true;
		}
	}
	
}
